### PR TITLE
b/253231544: Extract client IP from forward header

### PR DESCRIPTION
**Pull request by qiwzhang**
In some environment, client IP is set in `forward` header.  Define a flag to extract it from that header.

As the headers can be easily spoofed, this flag should only be used in a truth environment.

### CODE REVIEW

File: src/envoy/http/service_control/handler_utils.cc [hunk 1]
```diff
@@ -51,6 +54,8 @@ constexpr char kJwtPayLoadsDelimeter = '.';
 constexpr char kContentTypeApplicationGrpcPrefix[] = "application/grpc";
 const Envoy::Http::LowerCaseString kContentTypeHeader{"content-type"};
 
+const Envoy::Http::LowerCaseString kForwardHeader{"forward"};
```
[the remark this answers is not in the document]
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Thanks for catching this.   Done

File: src/envoy/http/service_control/handler_utils.cc [hunk 2]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
+
+  absl::string_view source = values[0]->value().getStringView();
+
+  // Multiple proxy sections are separated by comma. Use the last one.
+  const absl::string_view::size_type pos = source.find_last_of(',');
+  if (pos != absl::string_view::npos) {
+    source.remove_prefix(pos + 1);
+  }
+
+  // Each section may have multiple fields, they are separated by semicolumn.
```
Comment (KyleWiese) [inline src/envoy/http/service_control/handler_utils.cc:386]: nit: semicolon

File: tests/integration_test/service_control_apikey_restriction_test/service_control_apikey_restriction_test.go [hunk 3]
```diff
@@ -173,6 +176,62 @@ func TestServiceControlAPIKeyIpRestriction(t *testing.T) {
...
+			wantScRequest: &utils.ExpectedCheck{
+				Version:         utils.ESPv2Version(),
+				ServiceName:     "echo-api.endpoints.cloudesf-testing.cloud.goog",
+				ServiceConfigID: "test-config-id",
+				ConsumerID:      "api_key:api-key3",
+				OperationName:   "1.echo_api_endpoints_cloudesf_testing_cloud_goog.Echo",
+				ApiKey:          "api-key3",
+				CallerIp:        "1.2.3.4",
+			},
+		},
```
Comment (KyleWiese) [inline tests/integration_test/service_control_apikey_restriction_test/service_control_apikey_restriction_test.go:227]: May be worth adding a test around using forwarded even if x-forwarded-for is also specified?

**Review by TAOXUY — APPROVED** [review]

File: src/envoy/http/service_control/handler_utils.cc [hunk 4]
```diff
@@ -359,6 +364,54 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
 }
 
+std::string extractIPFromForwardHeader(
+    const Envoy::Http::RequestHeaderMap& headers) {
+  const auto values = headers.get(kForwardHeader);
+
+  // Only support one header value.
+  // Need to extract the last one,  could not define the last one for multiple header values.
+  if (values.size() != 1) {
+    return Envoy::EMPTY_STRING;
```
[the remark this answers is not in the document]
Comment (TAOXUY) [reply src/envoy/http/service_control/handler_utils.cc:375]: Why not log it in DEBUG?

File: src/envoy/http/service_control/handler_utils.cc [hunk 5]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
+  // header values.
+  if (values.size() != 1) {
+    return Envoy::EMPTY_STRING;
+  }
+
+  absl::string_view source = values[0]->value().getStringView();
+
+  // Multiple proxy sections are separated by comma. Use the last one.
+  const absl::string_view::size_type pos = source.find_last_of(',');
+  if (pos != absl::string_view::npos) {
```
Comment (nareddyt) [inline src/envoy/http/service_control/handler_utils.cc:385]: just return early if condition is true

File: src/envoy/http/service_control/handler_utils.cc [hunk 6]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
+    source.remove_prefix(pos + 1);
+  }
+
+  // Each section may have multiple fields, they are separated by semicolumn.
+  // For example: by=abc;for=1.2.3.4;host=abc;proto=http
+  // Extract ip from the "for=" field.
+  for (absl::string_view s : absl::StrSplit(source, ';')) {
```
Comment (nareddyt) [inline src/envoy/http/service_control/handler_utils.cc:392]: probably should extract out all magic strings and numbers into constants

File: src/envoy/http/service_control/handler_utils.cc [hunk 7]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
+    absl::string_view token = absl::StripAsciiWhitespace(s);
+    if (absl::StartsWith(token, "for=")) {
+      absl::string_view ip = token.substr(4);
```
Comment (nareddyt) [inline src/envoy/http/service_control/handler_utils.cc:395]: token.substr(kTokenPrefix.length())

File: src/envoy/http/service_control/handler_utils.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Done

File: src/envoy/http/service_control/handler_utils.cc [hunk 8]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
...
+        ip = ip.substr(1, ip.size() - 2);
+      }
+      // Remove [].
+      if (ip[0] == '[' && ip[ip.size() - 1] == ']') {
+        ip = ip.substr(1, ip.size() - 2);
+      }
+
+      // Verify it is a valid IP.
+      const auto address =
+          Envoy::Network::Utility::parseInternetAddressNoThrow(std::string(ip));
```
Comment (nareddyt) [inline src/envoy/http/service_control/handler_utils.cc:409]: move std::string construction outside function arg, use in return val below

File: src/envoy/http/service_control/handler_utils.cc [hunk 9]
```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
   info.grpc_response_code = static_cast<StatusCode>(status.value());
 }
 
+std::string extractIPFromForwardedHeader(
```
Comment (nareddyt) [inline src/envoy/http/service_control/handler_utils.cc]: add fuzz test for custom parser

File: src/envoy/http/service_control/handler_utils.cc (hunk 2 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Done

File: tests/integration_test/service_control_apikey_restriction_test/service_control_apikey_restriction_test.go (hunk 3 above)
Comment (qiwzhang) [reply tests/integration_test/service_control_apikey_restriction_test/service_control_apikey_restriction_test.go:246]: Done

File: src/envoy/http/service_control/handler_utils.cc (hunk 7 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Done

File: src/envoy/http/service_control/handler_utils.cc (hunk 6 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc:394]: Updated for "for=",  but not for single character which is more readable here.

File: src/envoy/http/service_control/handler_utils.cc (hunk 8 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Done

File: src/envoy/http/service_control/handler_utils.cc (hunk 9 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc]: Done

File: src/envoy/http/service_control/handler_utils.cc (hunk 5 above)
Comment (qiwzhang) [reply src/envoy/http/service_control/handler_utils.cc:387]: Not sure what you mean here?

**Pull request merged**